src/render/decorations/CHyprGroupBarDecoration.cpp

```diff
@@ -57,110 +51,156 @@ void CHyprGroupBarDecoration::updateWindow(CWindow* pWindow) {
 
-        const bool         GROUPLOCKED  = m_pWindow->getGroupHead()->m_sGroupData.locked;
-        const auto* const  PCOLACTIVE   = GROUPLOCKED ? PGROUPCOLACTIVELOCKED : PGROUPCOLACTIVE;
-        const auto* const  PCOLINACTIVE = GROUPLOCKED ? PGROUPCOLINACTIVELOCKED : PGROUPCOLINACTIVE;
+            if (backBox.width > 0 && backBox.height > 0)
+                g_pHyprOpenGL->renderRect(&backBox, color, ROUNDING * pMonitor->scale);
 
-        CColor             color =
+            // cleanup stencil
+            glClearStencil(0);
+            glClear(GL_STENCIL_BUFFER_BIT);
+            glDisable(GL_STENCIL_TEST);
+            glStencilMask(-1);
+            glStencilFunc(GL_ALWAYS, 1, 0xFF);
+        } else if (backBox.width > 0 && backBox.height > 0)
+            g_pHyprOpenGL->renderRect(&backBox, color, ROUNDING * pMonitor->scale);
+    }
+
+    for (int i = 0; i < barsToDraw; ++i) {
+
+        barBox.x  = pos.x + currentOffset;
+        gradBox.x = pos.x + currentOffset;
+        textBox.x = pos.x + currentOffset;
+
+        CColor color =
             m_dwGroupMembers[i] == g_pCompositor->m_pLastWindow ? ((CGradientValueData*)PCOLACTIVE->get())->m_vColors[0] : ((CGradientValueData*)PCOLINACTIVE->get())->m_vColors[0];
         color.a *= a;
-        g_pHyprOpenGL->renderRect(&rect, color);
+        if (barBox.width > 0 && barBox.height > 0)
+            g_pHyprOpenGL->renderRect(&barBox, color, std::sqrt(m_iBarHeight) / 2);
 
         // render title if necessary
         if (*PRENDERTITLES) {
             CTitleTex* pTitleTex = textureFromTitle(m_dwGroupMembers[i]->m_szTitle);
-
             if (!pTitleTex)
-                pTitleTex =
-                    m_sTitleTexs.titleTexs
-                        .emplace_back(std::make_unique<CTitleTex>(m_dwGroupMembers[i], Vector2D{BARW * pMonitor->scale, (*PTITLEFONTSIZE + 2 * BAR_TEXT_PAD) * pMonitor->scale}))
-                        .get();
-
-            rect.height = (*PTITLEFONTSIZE + 2 * BAR_TEXT_PAD) * 0.8 * pMonitor->scale;
-
-            rect.y -= rect.height;
-            rect.width = BARW * pMonitor->scale;
-
-            refreshGradients();
+                pTitleTex = m_sTitleTexs.titleTexs.emplace_back(std::make_unique<CTitleTex>(m_dwGroupMembers[i], Vector2D(textBox.width, textBox.height))).get();
 
-            if (*PGRADIENTS)
-                g_pHyprOpenGL->renderTexture((m_dwGroupMembers[i] == g_pCompositor->m_pLastWindow ? m_tGradientActive : m_tGradientInactive), &rect, 1.0);
+            if (textBox.width > 0 && textBox.height > 0)
+                g_pHyprOpenGL->renderTexture(pTitleTex->tex, &textBox, 1.f);
+        }
 
-            rect.y -= (*PTITLEFONTSIZE + 2 * BAR_TEXT_PAD) * 0.2 * pMonitor->scale;
-            rect.height = (*PTITLEFONTSIZE + 2 * BAR_TEXT_PAD) * pMonitor->scale;
+        if (gradBox.width > 0 && gradBox.height > 0) {
+            if (m_tGradientActive.m_iTexID == 0) // no idea why it doesn't work
```

Comment:
when removing this, somehow the `refreshGradients()` function gets called, but texture doesn't exist
I haven't yet found what I'm doing wrong